theano/sandbox/gpuarray/opt.py

```diff
@@ -167,20 +237,39 @@ def local_gpuaalloc2(node):
 
     Moves an alloc that is an input to join to the gpu.
     """
-    if (isinstance(node.op, tensor.Alloc) and
-        all(c != 'output' and
-            c.op == tensor.join and
-            all(i.owner and
-                i.owner.op in [host_from_gpu, tensor.alloc]
-                for i in c.inputs[1:])
-            for c, idx in node.outputs[0].clients)):
-        return [host_from_gpu(gpu_alloc(*node.inputs))]
+    if isinstance(node.op, tensor.Alloc):
+        replace = True
+        # We try to match an existing context if there is one
+        context = None
+        for c, idx in node.outputs[0].clients:
+            if c == 'output':
+                replace = False
+            elif c.op == tensor.join:
+                for i in c.inputs[1:]:
+                    if not i.owner:
+                        replace = False
+                    if i.owner.op == host_from_gpu:
+                        context = i.owner.inputs[0].type.context
```

Comment:
First or last does not matter much, it's a stupid way to choose.  And it would only cause problems for nodes which have conflicting contexts.